### PR TITLE
RLlib + SMAC Example

### PR DESCRIPTION
**Pull request by richardliaw**

### CODE REVIEW

**Comment by richardliaw** [comment]
cc @ericl

File: smac/env/starcraft2/starcraft2.py [hunk 1]
```diff
@@ -1097,7 +1097,7 @@ def kill_all_units(self):
         debug_command = [
             d_pb.DebugCommand(kill_unit=d_pb.DebugKillUnit(tag=units_alive))
         ]
-        self.controller.debug(debug_command)
+#        self.controller.debug(debug_command)
```
Comment (ericl) [inline smac/env/starcraft2/starcraft2.py]: The environment seemed to be crashing during reset() if I left this in.

**Comment by ericl** [comment]
Updated to break the example into separate files for clarity. I think this is probably good to merge; perf should presumably be similar to https://github.com/ray-project/ray/pull/3542 since the code is largely unchanged.

File: smac/env/starcraft2/starcraft2.py (hunk 1 above)
Comment (samvelyan) [reply smac/env/starcraft2/starcraft2.py]: This error indicates that you're using the latest release of PySC2 which does not include the debug command in the remote controller. Use the master branch of PySC2 from the Github repository. `pip install git+https://github.com/deepmind/pysc2@a9f093493c4c77adb385602790a480e7f238b97d`

**Comment by samvelyan** [comment]
It seems like the total number of SC2 processes created during either `run_ppo` or `run_qmix` scripts is `num-workers + 1`, where one of the environments is not being used but takes extra memory. 
Any idea why this happens? 

Could please also state the additional dependencies in the README file, e.g. gym, tf?

**Comment by ericl** [comment]
Ah, the reason we create n+1 envs is that one is needed for the driver process. It seems like we need to create the StarCraft2 env there to get its action and observation space. However we never call any methods ok the env.

Is it possible to defer creation of the StarCraft process until reset() is called? That would allow access to obs and action state without extra overheads.

**Comment by ericl** [comment]
Added TF dep in readme. The rllib package already includes the rest of the deps.

**Comment by samvelyan** [comment]
Done in 91761f1

**Comment by ericl** [comment]
Awesome, I think that should fix the extra memory use issue then.

File: smac/env/starcraft2/starcraft2.py [hunk 2]
```diff
@@ -263,6 +264,9 @@ def __init__(
         self._sc2_proc = None
         self._controller = None
 
+        # Try to avoid leaking SC2 processes on shutdown
+        atexit.register(lambda: self.close())
```
Comment (ericl) [inline smac/env/starcraft2/starcraft2.py]: we could also do this in the RLlib wrapper, but I'm guessing killing the process on exit is probably desirable by default?

**Comment by ericl** [comment]
@samvelyan will be fixed in the next RLlib release https://github.com/ray-project/ray/pull/3810

File: smac/examples/rllib/env.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+                episode is just started, the value will be None.
+            dones (dict): Done values for each ready agent. The special key
+                "__all__" (required) is used to indicate env termination.
+            infos (dict): Optional info values for each agent id.
+        """
+
+        actions = []
+        for i in self._ready_agents:
+            if i not in action_dict:
+                raise ValueError(
```
Comment (xiaoToby) [inline smac/examples/rllib/env.py:73]: I met this valueError, how to fix it ?

**Pull request merged**